### PR TITLE
Starter of Pattern Matching in Rust

### PR DESCRIPTION
**Pull request by wch510719826**

### CODE REVIEW

File: Pattern Matching in Rust.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+It looks quite similar to the switch replica:
+
+```c++
+int Fib(int n) {
+	switch (n) {
+	case 1: return 1;
+	case 2: return 2;
+	default: return fib(n - 1) + fib(n - 2);
+	}
```
Comment (Vigilans) [inline Pattern Matching in Rust.md]: tab is used here

File: Pattern Matching in Rust.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+}
+```
+
+However, pattern matching is capable of type matching and binding, which works wonders when it comes to error handling. For example, it is quite common to check and retrieve from a collection in C++:
```
Comment (Vigilans) [inline Pattern Matching in Rust.md]:
```suggestion
However, `switch` statement only matches variables of integral type in C++ (often an enum class is used in place of  magic numbers). This is pretty primitive and may become verbose when more than one variable are needed to determine a dispatch.

But in Rust, a `match` statement can match variables of any type. With the combination of structual information and type system, Rust's pattern matching is able to deliver a neat, powerful, and type-safe control flow.

For example, it is quite common to check and retrieve from a collection in C++:
```

File: Pattern Matching in Rust.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    match table.get(&input) {
+        Some(val) => println!("{}", val),
+        None => println!("key {} doesn't exist", input),
+    }
+}
+```
+
+The result of the retrieval operation can properly be handled whether it is an error or a success. It doesn't have to access the object twice nor does it have to stuff the code with try-catch clauses.
+
+This is only the tip of the iceberg. When the control flow and data structure get more complicated, pattern matching will become more of a necessity than a quality of life change, as we'll demonstrate with a practical example in the next section.
```
Comment (Vigilans) [inline Pattern Matching in Rust.md]: This is a big topic of error handling in rust. I would rather suggest you to open a single section for it, and move the below message-dispatching code as the `for-example`.

Please check this thread:
https://stackoverflow.com/questions/42917566/what-is-this-question-mark-operator-about

You need to at least introduce the `?` operator here for the error handling.

Comment (wch510719826) [reply Pattern Matching in Rust.md]: Added chapter 'Pattern Matching and Error Handling'

**Pull request merged**